### PR TITLE
Developer documentation for Omega and bindings

### PR DESCRIPTION
**Pull request by jaylett**
Omega tests and contributing new bindings, plus some minor details.

### CODE REVIEW

File: contributing/new-bindings.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+Ambition
+--------
+
+We'd like Xapian to be usable from as many languages as possible. However,
+this means more than just having bindings code. To truly support a language,
+we need to have:
+
+* working bindings that can be built against a latest version of Xapian and the
+  latest version of the target language
```
Comment (ojwb) [inline contributing/new-bindings.rst]: It seems a bit odd to say "a latest version of Xapian", though I guess there are multiple latest versions (one per branch).

File: contributing/new-bindings.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+* documentation to help people familiar with the target language to start using
+  the bindings
+
+* example code -- ideally translations of the code in the `Xapian user guide`_.
+  You can start from the `user guide source code`_, which already supports a few
+  languages.
+
+* automated tests for the bindings; at least a "smoketest" hooked into the tests
+  for the build system, so we'll notice
```
Comment (ojwb) [inline contributing/new-bindings.rst]: Unfinished paragraph?

Comment (ojwb) [reply contributing/new-bindings.rst]: One thing probably worth noting here is that for languages wrapped via SWIG we don't currently aim to test every single wrapped method, but rather rely on the xapian-core test suite to test xapian-core's behaviour via the C++ API, along with the typemap-based nature of SWIG's wrapping.  If we test one method which takes a `const std::string&` parameter then that provides test coverage for the typemap which handles that and we gain very little over that by testing every method which takes `const std::string&`.

So it's particularly important to have coverage for custom typemaps, and for anything else where we customise what SWIG produces by default.  It's also good to cover the SWIG built-in typemaps, so we catch any changes there which affect us.

And we also want regression tests for any bugs fixed.

File: contributing/new-bindings.rst [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	There are some basic Go bindings written by Marius Tibeica in the
+	"golang" branch, which is currently based on the ``RELEASE/1.4`` branch.
+
+Ocaml
+	Dan Colish did `some initial work on Ocaml support <ocaml_>`_.
+
+
+Guile
+	rm@fabula.de did some work on getting Guile bindings working, but sadly
+	most of this was lost when his laptop's hard disk died.
```
Comment (ojwb) [inline contributing/new-bindings.rst]: I think we should probably just drop this note - it's not actually useful to anyone wanting to work on Guile bindings to know that somebody did some work many years ago but it was lost.

File: conventions/deprecation.rst [hunk 4]
```diff
@@ -65,3 +65,52 @@ You can simply mark a method defined inline in a class with
...
+     warn "[DEPRECATION] World::hi() is deprecated, use World::hello() instead"
+
+* Perl:
+
+  .. code-block:: perl
+
+     use warnings;
+     warnings::warnif('deprecated', 'World::hi() is deprecated, use World::hello() instead');
+
+* Java has ``@Deprecated``, but I think that's a documentation thing only.
```
Comment (ojwb) [inline conventions/deprecation.rst]: Actually, I was confusing `@deprecated` and `@Deprecated` here I think - the latter is an annotation and "ensures" the compiler will issue a warning - the former is javadoc and may give a warning with some compilers: https://docs.oracle.com/javase/7/docs/technotes/guides/javadoc/deprecation/deprecation.html#annotation

File: contributing/new-bindings.rst (hunk 1 above)
Comment (jaylett) [reply contributing/new-bindings.rst]: Yeah, I meant that if someone contributes against 1.4.x but it doesn't work with master, that's still useful. It's not clear wording though, so maybe it should be "working bindings that can be built against a recent release of Xapian or git master, as well as against the latest version of the target language".

**Comment by ojwb** [comment]
Thanks.  Rebased, autosquashed and pushed as cbed278a46760b2750c388c7f484cf1231ebefe3.

**Pull request closed without merging**